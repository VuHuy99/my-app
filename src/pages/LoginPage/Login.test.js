import * as React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import Login from "./LoginPage";
import { configureStore } from '@reduxjs/toolkit';
import '@testing-library/jest-dom'
// Mock matchMedia for Ant Design components that rely on it


const mockReducer = {
  provider: (state = { users: {}, authedUser: null }) => state, // Simple mock reducer
};

describe('LoginPage', () => {
  let storeMock;
  beforeAll(() => {
    // Mocking window.matchMedia to return a mock function
    global.matchMedia = global.matchMedia || function() {
      return {
        matches: false,
        addListener: jest.fn(),
        removeListener: jest.fn()
      };
    };
  });
  beforeEach(() => {
    // Initialize the mock store with a basic state structure
    storeMock = configureStore({
      reducer: mockReducer, // Pass the mock reducer
      preloadedState: { 
        provider: {
          users: {
            user1: { id: "user1", name: "User One" },
            user2: { id: "user2", name: "User Two" },
          },
          authedUser: { id: "user1" }, // Assuming `authedUser` is part of the state
        },
      },
    });
  });
  test('will match snapshot',  () => {
      // eslint-disable-next-line testing-library/render-result-naming-convention
      var component = render(
          <MemoryRouter>
              <Provider store={storeMock}>
              <Login/>
              </Provider>
          </MemoryRouter>
      ) ;
      expect(component).toMatchSnapshot();
  })
  test("renders login page and displays loading spinner initially", async () => {
    render(
      <MemoryRouter>
              <Provider store={storeMock}>
              <Login/>
              </Provider>
          </MemoryRouter>
    );

    // Đảm bảo không có spinner khi component được render hoàn tất
    await waitFor(() =>
      expect(screen.queryByRole("spinbutton")).not.toBeInTheDocument()
    );
  });

  test("should render the login page with the title and dropdown", () => {
    render(
      <Provider store={storeMock}>
        <MemoryRouter initialEntries={["/login"]}>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    // test title and dropdown
    expect(screen.getByText("Please sign in to continue")).toBeInTheDocument();
  });
  test("should display user options in the dropdown", async () => {
        render(
          <Provider store={storeMock}>
          <MemoryRouter initialEntries={["/login"]}>
            <Login />
          </MemoryRouter>
        </Provider>
        );
    
        // test dropdown
        const select = screen.getByText("Select a user");
        expect(select).toBeInTheDocument();
    
        fireEvent.mouseDown(select);
    
        await waitFor(() => {
          expect(screen.getByText("User One")).toBeInTheDocument();
        });
      });
     test("renders login page and displays loading spinner initially", async () => {
    render(
      <Provider store={storeMock}>
        <MemoryRouter initialEntries={["/login"]}>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() =>
      expect(screen.queryByRole("spinbutton")).not.toBeInTheDocument()
    );
  });
  test("should display user options in the dropdown", async () => {
    render(
      <Provider store={storeMock}>
        <MemoryRouter initialEntries={["/login"]}>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    // test dropdown
    const select = screen.getByText("Select a user");
    expect(select).toBeInTheDocument();

    fireEvent.mouseDown(select);

    await waitFor(() => {
      expect(screen.getByText("User One")).toBeInTheDocument();
    });
  });

})

 






