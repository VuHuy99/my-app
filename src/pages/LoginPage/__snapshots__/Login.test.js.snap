// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should display user options in the dropdown 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        align="center"
        class="Loginpage"
      >
        <div
          class="ant-row table css-dev-only-do-not-override-apn68"
          style="margin-left: -8px; margin-right: -8px; row-gap: 16px;"
        >
          <div
            class="ant-col ant-col-24 title css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <span
              class="title-content"
            >
              Welcome to the Would You Rather App!
            </span>
            <span>
              Please sign in to continue
            </span>
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <img
              alt="Avatar"
              class="imageAvatars"
              src="/images/avatars/avatars.jpg"
            />
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <span
              style="color: rgb(255, 102, 0); font-weight: bold; font-size: 18px;"
            >
              Sign In
            </span>
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <form
              class="form"
            >
              <div>
                <div
                  class="ant-select ant-select-outlined css-dev-only-do-not-override-apn68 ant-select-single ant-select-show-arrow ant-select-open"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                        aria-autocomplete="list"
                        aria-controls="rc_select_TEST_OR_SSR_list"
                        aria-expanded="true"
                        aria-haspopup="listbox"
                        aria-owns="rc_select_TEST_OR_SSR_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="rc_select_TEST_OR_SSR"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    >
                      Select a user
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
              <div>
                <button
                  class="ant-btn css-dev-only-do-not-override-apn68 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  disabled=""
                  type="button"
                >
                  <span>
                    Login
                  </span>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="ant-select-dropdown ant-slide-up-appear ant-slide-up-appear-prepare ant-slide-up css-dev-only-do-not-override-apn68 ant-select-dropdown-placement-bottomLeft"
        style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
      >
        <div>
          <div
            id="rc_select_TEST_OR_SSR_list"
            role="listbox"
            style="height: 0px; width: 0px; overflow: hidden;"
          >
            <div
              aria-label="User One"
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_0"
              role="option"
            >
              user1
            </div>
            <div
              aria-label="User Two"
              aria-selected="false"
              id="rc_select_TEST_OR_SSR_list_1"
              role="option"
            >
              user2
            </div>
          </div>
          <div
            class="rc-virtual-list"
            style="position: relative;"
          >
            <div
              class="rc-virtual-list-holder"
              style="max-height: 256px; overflow-y: hidden;"
            >
              <div>
                <div
                  class="rc-virtual-list-holder-inner"
                  style="display: flex; flex-direction: column;"
                >
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option ant-select-item-option-active"
                    title="User One"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      User One
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                  <div
                    aria-selected="false"
                    class="ant-select-item ant-select-item-option"
                    title="User Two"
                  >
                    <div
                      class="ant-select-item-option-content"
                    >
                      User Two
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-item-option-state"
                      style="user-select: none;"
                      unselectable="on"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      align="center"
      class="Loginpage"
    >
      <div
        class="ant-row table css-dev-only-do-not-override-apn68"
        style="margin-left: -8px; margin-right: -8px; row-gap: 16px;"
      >
        <div
          class="ant-col ant-col-24 title css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <span
            class="title-content"
          >
            Welcome to the Would You Rather App!
          </span>
          <span>
            Please sign in to continue
          </span>
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <img
            alt="Avatar"
            class="imageAvatars"
            src="/images/avatars/avatars.jpg"
          />
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <span
            style="color: rgb(255, 102, 0); font-weight: bold; font-size: 18px;"
          >
            Sign In
          </span>
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <form
            class="form"
          >
            <div>
              <div
                class="ant-select ant-select-outlined css-dev-only-do-not-override-apn68 ant-select-single ant-select-show-arrow ant-select-open"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="true"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Select a user
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
            <div>
              <button
                class="ant-btn css-dev-only-do-not-override-apn68 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                disabled=""
                type="button"
              >
                <span>
                  Login
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`LoginPage will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        align="center"
        class="Loginpage"
      >
        <div
          class="ant-row table css-dev-only-do-not-override-apn68"
          style="margin-left: -8px; margin-right: -8px; row-gap: 16px;"
        >
          <div
            class="ant-col ant-col-24 title css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <span
              class="title-content"
            >
              Welcome to the Would You Rather App!
            </span>
            <span>
              Please sign in to continue
            </span>
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <img
              alt="Avatar"
              class="imageAvatars"
              src="/images/avatars/avatars.jpg"
            />
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <span
              style="color: rgb(255, 102, 0); font-weight: bold; font-size: 18px;"
            >
              Sign In
            </span>
          </div>
          <div
            class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
            style="padding-left: 8px; padding-right: 8px;"
          >
            <form
              class="form"
            >
              <div>
                <div
                  class="ant-select ant-select-outlined css-dev-only-do-not-override-apn68 ant-select-single ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-autocomplete="list"
                        aria-controls="rc_select_TEST_OR_SSR_list"
                        aria-expanded="false"
                        aria-haspopup="listbox"
                        aria-owns="rc_select_TEST_OR_SSR_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="rc_select_TEST_OR_SSR"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    >
                      Select a user
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
              <div>
                <button
                  class="ant-btn css-dev-only-do-not-override-apn68 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                  disabled=""
                  type="button"
                >
                  <span>
                    Login
                  </span>
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      align="center"
      class="Loginpage"
    >
      <div
        class="ant-row table css-dev-only-do-not-override-apn68"
        style="margin-left: -8px; margin-right: -8px; row-gap: 16px;"
      >
        <div
          class="ant-col ant-col-24 title css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <span
            class="title-content"
          >
            Welcome to the Would You Rather App!
          </span>
          <span>
            Please sign in to continue
          </span>
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <img
            alt="Avatar"
            class="imageAvatars"
            src="/images/avatars/avatars.jpg"
          />
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <span
            style="color: rgb(255, 102, 0); font-weight: bold; font-size: 18px;"
          >
            Sign In
          </span>
        </div>
        <div
          class="ant-col ant-col-24 css-dev-only-do-not-override-apn68"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <form
            class="form"
          >
            <div>
              <div
                class="ant-select ant-select-outlined css-dev-only-do-not-override-apn68 ant-select-single ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Select a user
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
            <div>
              <button
                class="ant-btn css-dev-only-do-not-override-apn68 ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                disabled=""
                type="button"
              >
                <span>
                  Login
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
